sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java

```diff
@@ -613,8 +613,10 @@ public Symbol visitArrayComparisonExpression(ArrayComparisonExpression node, Exp
             Symbol leftSymbol = process(node.getLeft(), context);
             Symbol arraySymbol = process(node.getRight(), context);
 
+            if (arraySymbol.valueType().equals(DataTypes.UNDEFINED)) {
+                arraySymbol = cast(arraySymbol, new ArrayType(leftSymbol.valueType()));
```

Comment:
This cast is actually not necessary if we remove the CollectionType checks, because undefined types will be converted automatically during `FuncParams` matching. (I just ran your tests with the CollectionType casts removed).